### ISSUE TITLE
Introduce mechanism for terminating Dynamic Tests early

### ISSUE THREAD

**Issue by mmerdes** [junit-team/junit5#431]
Dynamic Tests are non-deterministic in nature. Sometimes it might not be clear to the test author (let alone the IDE) when or even if the execution will terminate. 
It has been suggested to introduce a timeout to stop the execution before it would terminate by itself. An alternative would be to terminate the execution when a certain number of failed (or successful tests)  have been executed.

**Comment by sormuras** [comment]
With #838 in the works, it is easy to get and store the initial time instant when a "@TestFactory"-annotated methods started. The user-supplied code then could calculate the difference to now and throw an (time out) exception in a "`requiredTest`" executable. That breaks the execution loop.

Or, a time-out value could be passed at the "@TestFactory" annotaion level and the interpreted internally.

**Comment by sormuras** [comment]
Or the user may supply a predicate object that is asked before every dynamic test execution whether the execution loop should continue or not.

**Comment by sormuras** [comment]
The time-out use-case is already achievable via:
```
@TestFactory
Stream<DynamicNode> dynamicTestsWithContainers() {
    Instant start = Instant.now();
    return Stream.of("A", "B", "C")
      .map(input -> dynamicContainer("Container " + input, Stream.of(
        dynamicTest("not null", () -> assertNotNull(input)),
        requiredTest("check duration",
            () -> {
              long duration = Duration.between(start, Instant.now()).toMillis();
              if (duration > 23) {
                fail("Time is over: " + duration + " ms");
              }
            }
        ),
        dynamicContainer("properties", Stream.of(
          dynamicTest("length > 0", () -> assertTrue(input.length() > 0)),
          dynamicTest("not empty", () -> assertFalse(input.isEmpty()))
        ))
      )));
}
```

**Comment by sormuras** [comment]
tl;dr #838 will solve this issue.

**Comment by sbrannen** [comment]
Using a "required test" that throws an exception seems to be a hack to me.

I definitely would not call that intuitive in terms of the programming model.

Similarly, consider what the user's code would look like when implementing a custom `Iterable` instead of returning a stream. At some point the iterable would have to return a "required test" that only throws an exception, and that just seems wrong to me.

IMHO, I believe we need a more elegant, first-class solution to the problem addressed by this issue -- one that works even if there is no such thing as a "required test".

**Comment by sbrannen** [comment]
Supporting a user-defined timeout per test factory and/or a predicate that is provided a contextual object (containing information such as number of successful test, number of failed tests, number of aborted tests, number of skipped tests, duration of execution thus far, etc.) would be more intuitive.

**Comment by sormuras** [comment]
The current iteration of #838 supports a `DynamicTest.dynamicTest()` overload that takes an user-provided `Predicat<Boolean>` that is evaluated after the executable ran. The `Boolean` parameter can easily replaced with the contextual object @sbrannen mentions above.

**Comment by nicolaiparlog** [comment]
As far as I know the dynamic test API makes no promise about when the tests will be executed - it is possible to change the implementation so that the factory method is called and the dynamic tests are executed much later. I think this is a good thing and maybe should even be documented.

If you share that view, any implementation of this feature should make sure to not accidentally rely on the immediacy of the test execution, thus locking JUnit into that behavior. (I realized all of this when looking at your idea for [how to implement the time out](https://github.com/junit-team/junit5/issues/431#issuecomment-299667418).)

**Comment by sormuras** [comment]
Too many concerns and open ends with #838 

 _"we should think it over more and brainstorm about the user experience and programming model before doing any actual coding. In other words, we need to design the feature first."_

**Comment by dmitry-timofeev** [comment]
I bumped into this issue when attempted to build a [prototype](https://github.com/pitest/pitest-junit5-plugin/pull/15) of early termination of tests in a container once a failed one is discovered in pitest-junit5-plugin. 

[Pitest](http://pitest.org/), a mutation coverage tool, runs tests against a mutation of the code they cover to determine if these tests are good enough — if they detect mutants. When it runs a parameterized or dynamic JUnit 5 test and detects a test failure, there is usually no need to run subsequent tests — but that is controlled by Pitest configuration.

The workaround I used involved an extension implementing `ExecutionCondition` that filters tests once a failure is detected. But as Pitest controls if an individual test failure must abort test execution, and JUnit Jupiter Engine instantiates the extensions, the extension has to use a global static object to communicate with the rest of the system. Can't say if it is a generally good idea, but in the case of that plugin a feature allowing to abort the execution of subsequent tests at [_platform_](https://github.com/pitest/pitest-junit5-plugin/blob/e1962b237340a8b441f05db524a4a53f7968e265/src/main/java/org/pitest/junit5/JUnit5TestUnit.java#L49) level seems to be useful, because no extra extensions and communication through globals would be needed.

**Comment by Captain-P-Goldfish** [comment]
I would like to have this issue reopened. 
I think I got a pretty good usecase for aborting the execution of dynamic tests:

I am currently writing selenium integration tests and I am using DynamicTests for this purpose to create a logical dynamic workflow in the test. The problem here is if even one of the steps fails it is to be assumed that the following tests will fail too. So it would be great if the `@TestFactory` annotation could be extended by an additional parameter `abortOnFail` or something similiar that stops the execution of additional tests if somewhere along the road a test failed
